File: tests/test_watchdog_ip.py
import unittest

import simpy

from ip_model_automation.watchdog_ip import WatchdogIpModel

FAST = {
    "reset_latency": 1,
    "arm_latency": 1,
    "kick_latency": 1,
    "disarm_latency": 1,
    "countdown_tick_latency": 1,
    "expire_latency": 1,
    "accept_config_latency": 1,
    "apply_timeout_latency": 1,
}


def make_model(env, **overrides):
    kwargs = dict(FAST, log_level="CRITICAL")
    kwargs.update(overrides)
    return WatchdogIpModel(env, **kwargs)


class TestWatchdogIpModel(unittest.TestCase):
    def test_kick_before_expiry_prevents_expiration(self):
        """template test_scenarios[0]: kick_before_expiry_prevents_expiration.

        Sustained kicking, one per cycle, against a 5-cycle timeout must never
        let the countdown reach zero. This is also the regression test for the
        SimPy Store.get() request-leak this dogfood round found: racing a fresh
        get() against a timeout every loop iteration without keeping the losing
        request alive let a later put() (kick()) be absorbed FIFO by an earlier
        abandoned getter instead of the live one, so kicks_received silently
        undercounted and the watchdog could still expire despite being kicked.
        """
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=5)

        def driver():
            yield model.arm()
            for _ in range(20):
                yield env.timeout(1)
                yield model.kick()

        env.process(driver())
        env.run(until=25)

        self.assertEqual(model.get_metrics()["kicks_received"], 20)
        self.assertEqual(model.get_metrics()["expirations"], 0)
        self.assertEqual(model.fsm_state["watchdog_main"], "ARMED")

    def test_countdown_reaches_zero_without_kick_expires(self):
        """template test_scenarios[1]: countdown_reaches_zero_without_kick_expires.

        Also the regression test for the second leak this dogfood round found:
        when the countdown expires, _run_armed()'s still-live, never-consumed
        get() request was abandoned without being cancelled, so it lingered on
        the Store ahead of _run_expired()'s own fresh get() -- a later disarm()
        was then absorbed by that stale request instead of clearing the latch.
        """
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=5)

        def driver():
            yield model.arm()
            yield env.timeout(20)  # never kicked -- countdown must reach zero
            yield model.disarm()

        env.process(driver())
        env.run(until=30)

        self.assertEqual(model.get_metrics()["expirations"], 1)
        self.assertEqual(model.get_metrics()["disarm_count"], 1)
        self.assertEqual(model.fsm_state["watchdog_main"], "DISARMED", "disarm did not clear the expired latch")

    def test_expired_latch_blocks_rearm_until_disarmed(self):
        """EXPIRED -> DISARMED is the only declared exit; ARM/KICK while expired
        have no declared transition and must have no effect (DLD 6.1)."""
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=3)

        def driver():
            yield model.arm()
            yield env.timeout(10)  # expires
            yield model.arm()  # ignored while EXPIRED
            yield model.kick()  # ignored while EXPIRED

        env.process(driver())
        env.run(until=15)

        self.assertEqual(model.fsm_state["watchdog_main"], "EXPIRED")
        self.assertEqual(model.get_metrics()["expirations"], 1)
        self.assertEqual(model.get_metrics()["arm_count"], 1, "the ignored re-arm while EXPIRED must not count")
        self.assertEqual(model.get_metrics()["kicks_received"], 0, "the ignored kick while EXPIRED must not count")

    def test_kick_and_disarm_have_no_effect_while_disarmed(self):
        """No declared transition from DISARMED on KICK/DISARM (DLD 6.1): there
        is nothing to kick or cancel yet, so both are accepted with no effect."""
        env = simpy.Environment()
        model = make_model(env)

        def driver():
            yield model.kick()
            yield model.disarm()

        env.process(driver())
        env.run(until=5)

        self.assertEqual(model.fsm_state["watchdog_main"], "DISARMED")
        self.assertEqual(model.get_metrics()["kicks_received"], 0)
        self.assertEqual(model.get_metrics()["disarm_count"], 0)

    def test_arm_has_no_effect_while_already_armed(self):
        """ARMED -> ARMED has no declared transition on ARM (only on KICK); a
        redundant ARM while already armed must not reload the countdown or
        count as a second arm (DLD 6.1)."""
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=5)
        captured = []

        def driver():
            yield model.arm()
            captured.append(model.countdown_cycles)  # sampled right at arm, before any tick decrements it
            yield env.timeout(2)  # let a couple of ticks decrement the countdown below 5
            yield model.arm()  # redundant, ignored -- must not reload the countdown back to 5
            captured.append(model.countdown_cycles)
            yield model.disarm()

        env.process(driver())
        env.run(until=10)

        self.assertEqual(model.fsm_state["watchdog_main"], "DISARMED")
        self.assertEqual(model.get_metrics()["arm_count"], 1, "the redundant ARM must not count as a second arm")
        self.assertEqual(captured[0], 5)
        self.assertLess(captured[1], 5, "the redundant ARM must not reload the countdown")

    def test_configure_timeout_flows_independently_of_countdown(self):
        """template test_scenarios[2]: configure_timeout_flows_independently_of_countdown.

        A CONFIGURE_TIMEOUT applied while ARMED must not disturb the running
        countdown -- it only takes effect the next time the watchdog arms.
        """
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=5)

        def driver():
            yield model.arm()
            yield env.timeout(2)
            yield model.configure_timeout(8)
            # the running countdown (5, now at 3 after 2 ticks) is unaffected:
            # 3 more cycles brings it to zero and expiry, not 8.
            yield env.timeout(10)

        env.process(driver())
        env.run(until=20)

        self.assertEqual(model.fsm_state["config_intake"], "IDLE")
        self.assertEqual(model.configured_timeout_cycles, 8)
        self.assertEqual(
            model.get_metrics()["expirations"], 1, "the stored config must not have reset the running countdown"
        )

        # the stored value does take effect on the *next* arm
        env2 = simpy.Environment()
        model2 = make_model(env2, default_timeout_cycles=5)

        captured = []

        def driver2():
            yield model2.configure_timeout(2)
            yield model2.arm()
            # sampled the instant ARM takes effect, before the next countdown
            # tick can decrement it
            captured.append(model2.countdown_cycles)

        env2.process(driver2())
        env2.run(until=10)
        self.assertEqual(captured, [2], "the newly configured timeout must load on arm")

    def test_reset_clears_state_before_first_command(self):
        env = simpy.Environment()
        model = make_model(env)
        self.assertEqual(model.fsm_state["watchdog_main"], "RESET")
        env.run(until=2)
        self.assertEqual(model.fsm_state["watchdog_main"], "DISARMED")

    def test_multiple_arm_disarm_cycles_leave_no_stale_getters(self):
        """Regression test for the request-leak bug class: repeated arm/kick/
        disarm and arm/expire/disarm cycles must never let a stale pending
        getter accumulate on the host command Store."""
        env = simpy.Environment()
        model = make_model(env, default_timeout_cycles=4)

        def driver():
            yield model.arm()
            yield env.timeout(1)
            yield model.kick()
            yield env.timeout(1)
            yield model.disarm()

            yield model.arm()
            yield env.timeout(10)  # expires
            yield model.disarm()

            yield model.arm()
            for _ in range(6):
                yield env.timeout(1)
                yield model.kick()
            yield model.disarm()

        env.process(driver())
        env.run(until=60)

        self.assertEqual(model.fsm_state["watchdog_main"], "DISARMED")
        self.assertEqual(model.get_metrics()["arm_count"], 3)
        self.assertEqual(model.get_metrics()["disarm_count"], 3)
        self.assertEqual(model.get_metrics()["expirations"], 1)
        self.assertEqual(model.get_metrics()["kicks_received"], 7)
        self.assertEqual(
            len(model._host_q.get_queue), 1, "exactly one live getter should remain -- the blocked DISARMED wait"
        )


if __name__ == "__main__":
    unittest.main()

File: systemc/tests/test_watchdog_ip.cpp
// Real, compiled SystemC testbench for WatchdogIpModel, validated against
// templates/watchdog_ip.template.yaml's test_scenarios -- the SystemC
// counterpart to tests/test_watchdog_ip.py, run independently (no shared
// process/state with the Python suite; each backend is validated for real on
// its own toolchain).
//
// SystemC's kernel is elaborated once per process, so every scenario below
// gets its own WatchdogIpModel instance and its own driver SC_THREAD, all
// constructed before the single sc_start() call and running concurrently
// within one shared simulated timeline -- not one process per scenario, and
// not scenarios sharing a model instance.

#include <systemc.h>

#include <iostream>
#include <string>
#include <vector>

#include "../models/watchdog_ip.h"

struct TestResult {
    std::string name;
    bool passed = false;
    std::vector<std::string> failures;
};

inline std::string stringify(const std::string& value) { return value; }
inline std::string stringify(long value) { return std::to_string(value); }
inline std::string stringify(int value) { return std::to_string(value); }
inline std::string stringify(double value) { return std::to_string(value); }

#define CHECK_EQ(result, actual, expected, msg)                                                            \
    do {                                                                                                    \
        auto _actual = (actual);                                                                            \
        auto _expected = (expected);                                                                        \
        if (!(_actual == _expected)) {                                                                      \
            (result).failures.push_back(                                                                    \
                std::string(msg) + ": expected " + stringify(_expected) + " got " + stringify(_actual)      \
            );                                                                                               \
        }                                                                                                    \
    } while (0)

// ------------------------------------------------------------------------ //
// template test_scenarios[0]: kick_before_expiry_prevents_expiration.
//
// Also the SystemC-side regression check for the request-timing bug the
// SimPy port's dogfooding found: a design that only re-checked its command
// queue on an *edge-triggered* event fire (rather than re-checking queue
// state before every wait()) would let a second queued command sit unhandled
// for up to a full tick. Sustained kicking here would still eventually
// expire under that bug, the same symptom the SimPy fix's regression test
// targets.
// ------------------------------------------------------------------------ //
SC_MODULE(KickPreventsExpiryScenario) {
    WatchdogIpModel watchdog;
    TestResult result;

    SC_CTOR(KickPreventsExpiryScenario) : watchdog("watchdog") {
        result.name = "kick_before_expiry_prevents_expiration";
        SC_THREAD(drive);
    }

    void drive() {
        watchdog.configure_timeout(20);
        watchdog.arm();
        for (int i = 0; i < 20; i++) {
            wait(1, SC_NS);
            watchdog.kick();
        }
        CHECK_EQ(result, watchdog.metrics["kicks_received"], 20L, "kicks_received");
        CHECK_EQ(result, watchdog.metrics["expirations"], 0L, "expirations");
        CHECK_EQ(result, watchdog.fsm_state["watchdog_main"], std::string("ARMED"), "fsm_state");
        result.passed = result.failures.empty();
    }
};

// ------------------------------------------------------------------------ //
// template test_scenarios[1]: countdown_reaches_zero_without_kick_expires.
// ------------------------------------------------------------------------ //
SC_MODULE(ExpiresThenDisarmScenario) {
    WatchdogIpModel watchdog;
    TestResult result;

    SC_CTOR(ExpiresThenDisarmScenario) : watchdog("watchdog") {
        result.name = "countdown_reaches_zero_without_kick_expires";
        SC_THREAD(drive);
    }

    void drive() {
        watchdog.configure_timeout(5);
        watchdog.arm();
        sc_time armed_at = sc_time_stamp();

        // Wait until expiry is observed via the fsm_state_changed event
        // (the SimPy test traces this with env.step(); SystemC's equivalent
        // is watching the model's own transition-notification event).
        while (watchdog.fsm_state["watchdog_main"] != "EXPIRED") {
            wait(watchdog.fsm_state_changed);
        }
        sc_time expired_at = sc_time_stamp();

        watchdog.disarm();

        CHECK_EQ(result, watchdog.metrics["expirations"], 1L, "expirations");
        CHECK_EQ(result, watchdog.metrics["disarm_count"], 1L, "disarm_count");
        CHECK_EQ(
            result, watchdog.fsm_state["watchdog_main"], std::string("DISARMED"),
            "disarm did not clear the expired latch"
        );
        // 5 countdown_tick cycles + the 1-cycle expire delay declared in the
        // template's watchdog_main.expire operation.
        CHECK_EQ(
            result, (expired_at - armed_at).to_default_time_units(), sc_time(6, SC_NS).to_default_time_units(),
            "expiry did not land at the declared countdown + expire delay"
        );
        result.passed = result.failures.empty();
    }
};

// ------------------------------------------------------------------------ //
// template test_scenarios[2]: configure_timeout_flows_independently_of_countdown.
// ------------------------------------------------------------------------ //
SC_MODULE(ConfigureTimeoutIndependentScenario) {
    WatchdogIpModel watchdog;
    TestResult result;

    SC_CTOR(ConfigureTimeoutIndependentScenario) : watchdog("watchdog", 1, 1, 1, 1, 1, 1, 1, 1, 5) {
        result.name = "configure_timeout_flows_independently_of_countdown";
        SC_THREAD(drive);
    }

    void drive() {
        // default_timeout_cycles=5 in the constructor above: the running
        // countdown (5, then 3 after 2 ticks) must reach zero and expire
        // well within this scenario's 10ns wait, unaffected by the later
        // configure_timeout(8) below.
        watchdog.arm();
        wait(2, SC_NS);
        sc_time requested_at = sc_time_stamp();
        watchdog.configure_timeout(8);
        sc_time apply_elapsed = sc_time_stamp() - requested_at;

        // The running countdown (5 default, now at 3 after 2 ticks) is
        // unaffected: 3 more cycles brings it to zero and expiry, not 8.
        wait(10, SC_NS);

        CHECK_EQ(result, watchdog.fsm_state["config_intake"], std::string("IDLE"), "config_intake fsm_state");
        CHECK_EQ(result, watchdog.configured_timeout_cycles, 8, "configured_timeout_cycles");
        CHECK_EQ(
            result, apply_elapsed.to_default_time_units(), sc_time(2, SC_NS).to_default_time_units(),
            "accept_config + apply_timeout did not take the declared 2 cycles"
        );
        CHECK_EQ(
            result, watchdog.metrics["expirations"], 1L, "the stored config must not have reset the running countdown"
        );
        result.passed = result.failures.empty();
    }
};

// ------------------------------------------------------------------------ //
// EXPIRED -> DISARMED is the only declared exit; ARM/KICK while expired have
// no declared transition and must have no effect (DLD 6.1).
// ------------------------------------------------------------------------ //
SC_MODULE(NoEffectWhileExpiredScenario) {
    WatchdogIpModel watchdog;
    TestResult result;

    SC_CTOR(NoEffectWhileExpiredScenario) : watchdog("watchdog") {
        result.name = "no_effect_while_expired";
        SC_THREAD(drive);
    }

    void drive() {
        watchdog.configure_timeout(3);
        watchdog.arm();
        while (watchdog.fsm_state["watchdog_main"] != "EXPIRED") {
            wait(watchdog.fsm_state_changed);
        }
        watchdog.arm();   // ignored while EXPIRED
        watchdog.kick();  // ignored while EXPIRED

        CHECK_EQ(result, watchdog.fsm_state["watchdog_main"], std::string("EXPIRED"), "fsm_state");
        CHECK_EQ(result, watchdog.metrics["expirations"], 1L, "expirations");
        CHECK_EQ(result, watchdog.metrics["arm_count"], 1L, "the ignored re-arm while EXPIRED must not count");
        CHECK_EQ(result, watchdog.metrics["kicks_received"], 0L, "the ignored kick while EXPIRED must not count");
        result.passed = result.failures.empty();
    }
};

// ------------------------------------------------------------------------ //
// No declared transition from DISARMED on KICK/DISARM (DLD 6.1): there is
// nothing to kick or cancel yet, so both are accepted with no effect.
// ------------------------------------------------------------------------ //
SC_MODULE(NoEffectWhileDisarmedScenario) {
    WatchdogIpModel watchdog;
    TestResult result;

    SC_CTOR(NoEffectWhileDisarmedScenario) : watchdog("watchdog") {
        result.name = "no_effect_while_disarmed";
        SC_THREAD(drive);
    }

    void drive() {
        watchdog.kick();
        watchdog.disarm();

        CHECK_EQ(result, watchdog.fsm_state["watchdog_main"], std::string("DISARMED"), "fsm_state");
        CHECK_EQ(result, watchdog.metrics["kicks_received"], 0L, "kicks_received");
        CHECK_EQ(result, watchdog.metrics["disarm_count"], 0L, "disarm_count");
        result.passed = result.failures.empty();
    }
};

// ------------------------------------------------------------------------ //
// ARMED -> ARMED has no declared transition on ARM (only on KICK); a
// redundant ARM while already armed must not reload the countdown or count
// as a second arm (DLD 6.1).
// ------------------------------------------------------------------------ //
SC_MODULE(NoEffectWhileArmedScenario) {
    WatchdogIpModel watchdog;
    TestResult result;
    int countdown_after_first_arm = -1;
    int countdown_after_redundant_arm = -1;

    SC_CTOR(NoEffectWhileArmedScenario) : watchdog("watchdog") {
        result.name = "no_effect_while_already_armed";
        SC_THREAD(drive);
    }

    void drive() {
        watchdog.configure_timeout(5);
        watchdog.arm();
        countdown_after_first_arm = watchdog.countdown_cycles;
        wait(2, SC_NS);  // let a couple of ticks decrement the countdown below 5
        watchdog.arm();  // redundant, ignored -- must not reload the countdown back to 5
        countdown_after_redundant_arm = watchdog.countdown_cycles;
        watchdog.disarm();

        CHECK_EQ(result, watchdog.fsm_state["watchdog_main"], std::string("DISARMED"), "fsm_state");
        CHECK_EQ(result, watchdog.metrics["arm_count"], 1L, "the redundant ARM must not count as a second arm");
        CHECK_EQ(result, countdown_after_first_arm, 5, "countdown_after_first_arm");
        if (countdown_after_redundant_arm >= 5) {
            result.failures.push_back("the redundant ARM must not reload the countdown");
        }
        result.passed = result.failures.empty();
    }
};

int sc_main(int argc, char* argv[]) {
    KickPreventsExpiryScenario s1("s1");
    ExpiresThenDisarmScenario s2("s2");
    ConfigureTimeoutIndependentScenario s3("s3");
    NoEffectWhileExpiredScenario s4("s4");
    NoEffectWhileDisarmedScenario s5("s5");
    NoEffectWhileArmedScenario s6("s6");

    sc_start(60, SC_NS);

    std::vector<TestResult*> results = {&s1.result, &s2.result, &s3.result, &s4.result, &s5.result, &s6.result};
    int failed = 0;
    for (auto* r : results) {
        if (r->passed) {
            std::cout << "PASS  " << r->name << std::endl;
        } else {
            failed++;
            std::cout << "FAIL  " << r->name << (r->failures.empty() ? " (did not complete)" : "") << std::endl;
            for (const auto& f : r->failures) {
                std::cout << "        " << f << std::endl;
            }
        }
    }
    std::cout << (static_cast<int>(results.size()) - failed) << "/" << results.size() << " passed" << std::endl;
    return failed == 0 ? 0 : 1;
}
